crates/matrix-sdk-common/src/deserialized_responses.rs

```diff
@@ -300,106 +300,79 @@ pub struct EncryptionInfo {
-    #[serde(skip_serializing_if = "Option::is_none")]
-    pub unsigned_encryption_info: Option<BTreeMap<UnsignedEventLocation, UnsignedDecryptionResult>>,
 }
 
 impl SyncTimelineEvent {
     /// Create a new `SyncTimelineEvent` from the given raw event.
     ///
-    /// This is a convenience constructor for when you don't need to set
-    /// `encryption_info` or `push_action`, for example inside a test.
+    /// This is a convenience constructor for a plaintext event when you don't
+    /// need to set `push_action`, for example inside a test.
     pub fn new(event: Raw<AnySyncTimelineEvent>) -> Self {
-        Self {
-            inner_event: event,
-            inner_encryption_info: None,
-            push_actions: vec![],
-            unsigned_encryption_info: None,
-        }
+        Self { inner: TimelineEventInner::PlainText { event }, push_actions: vec![] }
     }
 
     /// Create a new `SyncTimelineEvent` from the given raw event and push
     /// actions.
     ///
-    /// This is a convenience constructor for when you don't need to set
-    /// `encryption_info`, for example inside a test.
+    /// This is a convenience constructor for a plaintext event, for example
+    /// inside a test.
     pub fn new_with_push_actions(
         event: Raw<AnySyncTimelineEvent>,
         push_actions: Vec<Action>,
     ) -> Self {
-        Self {
-            inner_event: event,
-            inner_encryption_info: None,
-            push_actions,
-            unsigned_encryption_info: None,
-        }
+        Self { inner: TimelineEventInner::PlainText { event }, push_actions }
     }
 
     /// Get the event id of this `SyncTimelineEvent` if the event has any valid
     /// id.
     pub fn event_id(&self) -> Option<OwnedEventId> {
-        self.inner_event.get_field::<OwnedEventId>("event_id").ok().flatten()
+        self.inner.event().get_field::<OwnedEventId>("event_id").ok().flatten()
     }
 
     /// Returns a reference to the (potentially decrypted) Matrix event inside
     /// this `TimelineEvent`.
     pub fn event(&self) -> &Raw<AnySyncTimelineEvent> {
-        &self.inner_event
+        self.inner.event()
     }
 
     /// If the event was a decrypted event that was successfully decrypted, get
     /// its encryption info. Otherwise, `None`.
     pub fn encryption_info(&self) -> Option<&EncryptionInfo> {
-        self.inner_encryption_info.as_ref()
+        self.inner.encryption_info()
     }
 
     /// Takes ownership of this `TimelineEvent`, returning the (potentially
     /// decrypted) Matrix event within.
     pub fn into_event(self) -> Raw<AnySyncTimelineEvent> {
-        self.inner_event
+        self.inner.into_event()
     }
 
     /// Replace the Matrix event within this event. Used to handle redaction.
     pub fn replace_inner(&mut self, event: Raw<AnySyncTimelineEvent>) {
-        self.inner_event = event;
+        self.inner = TimelineEventInner::PlainText { event };
```

Comment:
I think it is correct, yes.

This case is only used in one very specific case: updating an existing event (which may or may not have been encrypted), when handling a redaction event. As you know, redaction removes all the content of an event, which means that, after redaction, it cannot possibly be successfully decrypted.

We could debate whether a redacted encrypted event is better classified as plaintext, unable-to-decrypt, or something else entirely, but this change preserves the existing functionality which is what I'm seeking to do in this PR.

Reply:
Wouldn't that result in messages being redacted appearing as insecure in an e2ee chat? (think it's already the case, fwiw, so if you're fine with this status quo for this PR and intend to change it later, let's do that)

I'm not satisfied with this invariant that it's only used for redaction, appearing only in the doc comment. Can we do one of these two things:

1. rename the function to `replace_as_redacted` and do something super specific? (maybe we don't even need the raw event anymore in this case?)
2. instead of this method, we provide `raw_mut(&mut self) -> &mut Raw<AnySyncTimelineEvent>` and the callers do whatever they want with the mutable ref? (including: replacing the inner event; but that's an explicit decision done at the caller's call-site)

Reply:
> Wouldn't that result in messages being redacted appearing as insecure in an e2ee chat? (think it's already the case, fwiw, so if you're fine with this status quo for this PR and intend to change it later, let's do that)

Yes.

> I'm not satisfied with this invariant that it's only used for redaction, appearing only in the doc comment. 

Fair. Having renamed `TimelineEventInner` to `TimelineEventKind`, we can just make `kind` public, and then the caller can replace `kind` directly, which makes the whole problem go away.

Reply:
https://github.com/matrix-org/matrix-rust-sdk/commit/ba762a895e284d03877b306b45d713dfdb3dd105 removes this, which solves the problem I think.